### PR TITLE
Porter cleanup

### PR DESCRIPTION
**Pull request by cavis**
Refactoring some of the Porter code, in prep for the "audio segmentation" workflow.  Which will involve copying files, generating wavforms, and slicing up files via ffmpeg.

Overview:

- `PorterEncoder` concern removed, and the pieces of it I still needed moved to `PorterUtils`
  - Reason being: the logic for what-tasks-to-send-to-porter and how-to-handle-porter-callbacks moved to the specific `Task` sub-models
  - This is much cleaner, for viewing what a Task asks Porter to do, and how it interprets that
- `PorterDecoder` concern renamed to `PorterCallback`
  - And as with above, the "how to interpret the callback JSON" falls to the `Task` sub-models
- `CopyMediaTask` and `CopyImageTask` now directly have their porter-input and parse-porter-callback logic.  All in 1 place!
- Tests shuffled a bit, and tried to get better coverage on some of these methods.

### CODE REVIEW

File: app/models/concerns/porter_utils.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+require "active_support/concern"
+
+module PorterUtils
```
Comment (cavis) [inline app/models/concerns/porter_utils.rb:3]: Chunks of this were copied from `PorterEncoder`.  But I reorganized the primary methods to reflect the structure of a [porter job](https://github.com/PRX/Porter#input-message-format).

Which has an ID, Source, Tasks, and Callbacks.

The common things (ID, Source, Callbacks) are defined here.  But the Tasks are up to the specific `Task` sub-model.

File: app/models/tasks/copy_image_task.rb [hunk 2]
```diff
@@ -1,46 +1,56 @@
...
+  def source_url
+    image.href
   end
 
-  def image_path(image_resource)
-    if !image_resource
-      logger.info("in CopyImageTask#image_path and image_resource is nil. self is #{inspect}")
-    end
-    URI.parse(image_resource.published_url).path
+  def porter_tasks
```
Comment (cavis) [inline app/models/tasks/copy_image_task.rb:16]: Much nicer to see the Porter input JSON all in one place!

File: app/models/tasks/copy_image_task.rb [hunk 3]
```diff
@@ -1,46 +1,56 @@
...
 
-  def podcast
-    image_resource.try(:episode).try(:podcast) || image_resource.try(:podcast)
-  end
+  def update_image
+    image.status = status
 
-  def image_resource
-    owner
+    if complete?
```
Comment (cavis) [inline app/models/tasks/copy_image_task.rb:38]: This code came from (1) the `before_save` block that was at the top of this file, and (2) the "parse callback metadata" logic from the old `PorterDecoder` concern.

File: test/models/task_test.rb [hunk 4]
```diff
@@ -2,37 +2,75 @@
 
 describe Task do
   let(:porter_task) { build(:porter_job_results) }
+  let(:task) { Task.new }
 
-  let(:task) { Task.create }
+  describe ".callback" do
```
Comment (cavis) [inline test/models/task_test.rb:7]: Trying to more consistently write `describe ".class_method"` and `describe "#instance_method"` syntax.

File: app/models/concerns/image_file.rb [hunk 5]
```diff
@@ -67,7 +67,7 @@ def file_name
   end
 
   def copy_media(force = false)
-    if !task || force
+    if force || !(status_complete? || task)
```
[the remark this answers is not in the document]
Comment (radical-ube) [reply app/models/concerns/image_file.rb:70]: what is the context for `copy_media`? and/or maybe an example of why files do/don't need to be copied, and that relationship to a Task (what is an example Task?)

File: app/models/concerns/porter_utils.rb (hunk 1 above)
Comment (radical-ube) [reply app/models/concerns/porter_utils.rb:3]: reading a bit through the Porter readme to better understand Tasks and Callbacks. i think a lot of my questions have been stemming from unfamiliarity with Porter and the workflow process between our apps and Porter.

File: app/models/task.rb [hunk 6]
```diff
@@ -39,29 +38,23 @@ def self.callback(msg)
...
+  def start!
+    self.status = "started"
+    self.options = {
+      Id: job_id,
+      Source: porter_source,
+      Tasks: porter_tasks,
+      Callbacks: porter_callbacks
+    }
+
+    porter_start!(options)
```
Comment (radical-ube) [inline app/models/task.rb:57]: i'm having trouble understanding this workflow. `porter_start!` starts a job, but i thought a job can contain tasks. so i'm a little confused on how this is being called from within a Task model/instance without causing an infinite call situation. unless jobs are started from Tasks?

Comment (cavis) [reply app/models/task.rb:57]: Terminology getting a bit overridden here... get ready for a long answer!

So in Feeder, we have a `Task` model in the DB.  Whose main purpose is to track a single [Porter Job](https://github.com/PRX/Porter#input-message-format).  And that Porter Job contains multiple Porter-Tasks.

So in the `CopyMediaTask`, we pull together the porter `job_id`, the `porter_source`, `porter_tasks` and `porter_callbacks` ... and assemble a chunk of JSON (saved to the db as `task.options`) which looks like this:

```json
{
  "Job": {
    "Id": "4ff6717f-a1e3-4f0f-a1a1-467a55026cce",
    "Source": {
      "Mode": "AWS/S3",
      "BucketName": "staging.mediajoint.prx.org",
      "ObjectKey": "public/audio_files/1212425/t121.mp3"
    },
    "Tasks": [
      {
        "Type": "Copy",
        "Mode": "AWS/S3",
        "BucketName": "staging-prx-feed",
        "ObjectKey": "878/4813ab85-902f-4eeb-967e-c39d7dc64c47/b5735ada-3662-4ea9-ad45-81ca5e53ee58.mp3",
        "ContentType": "REPLACE",
        "Parameters": {
          "CacheControl": "max-age=86400",
          "ContentDisposition": "attachment; filename=\"t121.mp3\""
        }
      },
      {
        "Type": "Inspect"
      }
    ],
    "Callbacks": [
      {
        "Type": "AWS/SQS",
        "Queue": "https://sqs.us-east-1.amazonaws.com/561178107736/4e489d22_stag_feeder_fixer_callback"
      }
    ]
  }
}
```

We also send that over [SNS](https://aws.amazon.com/sns/) (an AWS messaging thing ... like sending a text message) which triggers a [Porter execution](https://us-east-1.console.aws.amazon.com/states/home?region=us-east-1#/v2/executions/details/arn:aws:states:us-east-1:561178107736:execution:StateMachine-8B8z7vHLT4JS:a318d3b3-40cb-4f6c-9f6f-99d689ff015b).  (Porter is a "state machine" but that's maybe TMI for right now).  If you click on the "Execution input and output" tab you'll see that JSON.

So Porter performs those tasks (inspecting the source file, and copying it over to that location).  And then sends some callback JSON (you can also see it on the input-and-output tab) to the [AWS SQS](https://aws.amazon.com/sqs/) queue we indicated.

```json
{
  "Time": "2023-05-18T16:13:31.328Z",
  "Timestamp": 1684426411.328,
  "JobResult": {
    "Job": {
      "Id": "4ff6717f-a1e3-4f0f-a1a1-467a55026cce"
    },
    "Execution": {
      "Id": "arn:aws:states:us-east-1:561178107736:execution:StateMachine-8B8z7vHLT4JS:a318d3b3-40cb-4f6c-9f6f-99d689ff015b"
    },
    "FailedTasks": [],
    "State": "DONE",
    "TaskResults": [
      {
        "Task": "Copy",
        "Mode": "AWS/S3",
        "BucketName": "staging-prx-feed",
        "ObjectKey": "878/4813ab85-902f-4eeb-967e-c39d7dc64c47/b5735ada-3662-4ea9-ad45-81ca5e53ee58.mp3",
        "Time": "2023-05-18T16:13:30.206Z",
        "Timestamp": 1684426410.206
      },
      {
        "Task": "Inspect",
        "Inspection": {
          "Size": 482353,
          "Extension": "mp3",
          "MIME": "audio/mpeg",
          "Audio": {
            "Duration": 121051,
            "Format": "mp3",
            "Bitrate": 31878,
            "Frequency": 44100,
            "Channels": 2,
            "Layout": "stereo",
            "Layer": "3",
            "Frames": 4635
          }
        }
      }
    ]
  }
}
```

The Feeder worker is watching that SQS queue, sees the new message, picks it up, and calls [Task.callback](https://github.com/PRX/feeder.prx.org/blob/master/app/workers/porter_callback_worker.rb#L5) with the JSON.

That lands in the class method [self.callback](https://github.com/PRX/feeder.prx.org/blob/master/app/models/task.rb#L27) ... which tries to find a `Task` in the feeder db, with that Porter "Job Id" `4ff6717f-a1e3-4f0f-a1a1-467a55026cce`.

If found (and some other timestamp-check guards) ... we `task.update!(result: msg)](status: status, logged_at: time, result: msg)`.  Which drops us into the [before_save callback](https://github.com/PRX/feeder.prx.org/blob/master/app/models/tasks/copy_media_task.rb#L2)... which then parses that JSON result, pulling out the audio/image metadata.

Comment (radical-ube) [reply app/models/task.rb:57]: a little over my head but i think i kind of get the gist. a Task model in Feeder kind of represents a Porter Job, which itself consists of porter tasks, which are the individual tasks given to the Job by whichever Task model/instance defines those tasks, like `CopyImageTask` or `CopyMediaTask`.

**Review by radical-ube — APPROVED** [review]
i've read through to the best of my ability. i'm not sure how to QA any of this, but this PR doesn't seem to break the uploading process.

File: app/models/task.rb (hunk 6 above)
Comment (cavis) [reply app/models/task.rb:57]: +1 you got it.  And a lot more succinct than my description!

**Pull request merged**